Implement filtering for DOM events

Legioth opened the pull request:
The existing "event data" expressions feature is enhanced to also
include filter expressions, along with a boolean for each expression to
indicate whether that expressions should be used as a filter or not. The
value of filter expressions is sent back to the server in the same way
as regular data expressions, and then those are matched for true-ish
values when determining which listeners to notify.

This is an intermediate step for #468.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/flow/3912)
<!-- Reviewable:end -->

SomeoneToIgnore commented:
Reviewed 3 of 21 files at r1, 2 of 18 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1151 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANqpK4EBCh9Qcpjm4n:-LANqpK4EBCh9Qcpjm4o:bjtw1js) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1151)):*
> ```Java
>          *  1: at least one passing filter encountered
>          */
>         int filterStatus = 0;
> ```

I would have used `Boolean` with normal variable name, but up to you.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1159 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANq7S31qmdlFo0TEPs:-LANq7S31qmdlFo0TEPt:b7soe7o) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1159)):*
> ```Java
>             eventData = Json.createObject();
> 
>             for (String expressionString : expressions) {
> ```

Can you explain a bit about those expressions, where do they come from?

They aren't all filter expressions, right? Can there be non-boolean expression?
If so, how can we guarantee that when we do `expressionSettings.getBoolean(expressionString)` an exception won't be thrown?

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1178 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANl-xK-rJvd8dxf63A:-LANkojp8JCmDWiFEy0o:bdzp688) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1178)):*
> ```Java
>         }
> 
>         if (filterStatus != -1) {
> ```

Should we add a negative test case, where no filters match and no data is sent?

---

*[flow-server/src/main/java/com/vaadin/flow/dom/DomListenerRegistration.java, line 56 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANgt95DN26AxMo1-fE:-LANgt95DN26AxMo1-fF:bs8jug6) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-server/src/main/java/com/vaadin/flow/dom/DomListenerRegistration.java#L56)):*
> ```Java
> 
>     /**
>      * Sets a JavaScript expression that is used for filtering events to this
> ```

Up to you: might it make sense to mention that the expression should return `boolean` result more explicitly?

By the way, is it dangerous to evaluate generic JS strings here? Someone may put logging of critical information here or just annoying `alert`.

---

*[flow-server/src/main/java/com/vaadin/flow/internal/JsonUtils.java, line 266 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANaXVMD6xRt6JUMPDv:-LANaXVOE5zEDjnFBSRu:b-we7cy) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-server/src/main/java/com/vaadin/flow/internal/JsonUtils.java#L266)):*
> ```Java
>      * @return the created object
>      */
>     public static <T> JsonObject createObject(Map<String, T> map,
> ```

Should we add a test for this method also?

---

*[flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementListenerMap.java, line 59 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANtpp2EWunh8GXC-18:-LANtpp2EWunh8GXC-19:b-vkaszf) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementListenerMap.java#L59)):*
> ```Java
>         private DisabledUpdateMode mode;
>         private Set<String> eventDataExpressions;
>         private String filter = null;
> ```

It's `null` already, but whatever :)

---

*[flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/DomEventFilterView.java, line 32 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANcrdk2lSeAbXv7Iyi:-LANcrdk2lSeAbXv7Iyj:b-5wj4vm) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-tests/test-root-context/src/main/java/com/vaadin/flow/uitest/ui/DomEventFilterView.java#L32)):*
> ```
>  ==
> ```

Might it be more safe to use `===` here?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3912)*
<!-- Sent from Reviewable.io -->

Legioth commented:
Review status: 3 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1151 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANqpK4EBCh9Qcpjm4n:-LARS_sE5NSfn2K1XGUY:byril8r) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1151)):*
<details><summary><i>Previously, SomeoneToIgnore (Kirill Bulatov) wrote…</i></summary><blockquote>

I would have used `Boolean` with normal variable name, but up to you.
</blockquote></details>

Clarified even further by introducing two separate booleans. Also removed the short circuiting since it just made the code harder to read without any real-world performance benefit.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1159 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANq7S31qmdlFo0TEPs:-LARTU_15nrzKV3EBQk9:bwq18oo) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1159)):*
<details><summary><i>Previously, SomeoneToIgnore (Kirill Bulatov) wrote…</i></summary><blockquote>

Can you explain a bit about those expressions, where do they come from?

They aren't all filter expressions, right? Can there be non-boolean expression?
If so, how can we guarantee that when we do `expressionSettings.getBoolean(expressionString)` an exception won't be thrown?
</blockquote></details>

`getBoolean` checks for true-ish values according to JS semantics, so it doesn't throw for non-boolean values.

I added an overall description of the expression concept to the commit message.

---

*[flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java, line 1178 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANl-xK-rJvd8dxf63A:-LARUHtG58IjPbn6n2_C:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-client/src/main/java/com/vaadin/client/flow/binding/SimpleElementBindingStrategy.java#L1178)):*
<details><summary><i>Previously, SomeoneToIgnore (Kirill Bulatov) wrote…</i></summary><blockquote>

Should we add a negative test case, where no filters match and no data is sent?
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/flow/dom/DomListenerRegistration.java, line 56 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANgt95DN26AxMo1-fE:-LARVWjP3oDIUCNbOttZ:b-jpvwk9) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-server/src/main/java/com/vaadin/flow/dom/DomListenerRegistration.java#L56)):*
<details><summary><i>Previously, SomeoneToIgnore (Kirill Bulatov) wrote…</i></summary><blockquote>

Up to you: might it make sense to mention that the expression should return `boolean` result more explicitly?

By the way, is it dangerous to evaluate generic JS strings here? Someone may put logging of critical information here or just annoying `alert`.
</blockquote></details>

Documentation updated. This is a low level feature that may cause weird results if used in "irresponsible" ways.

---

*[flow-server/src/main/java/com/vaadin/flow/internal/JsonUtils.java, line 266 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANaXVMD6xRt6JUMPDv:-LAR_IpZALtQTp6g0E3h:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-server/src/main/java/com/vaadin/flow/internal/JsonUtils.java#L266)):*
<details><summary><i>Previously, SomeoneToIgnore (Kirill Bulatov) wrote…</i></summary><blockquote>

Should we add a test for this method also?
</blockquote></details>

Done.

---

*[flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementListenerMap.java, line 59 at r3](https://beta.reviewable.io/reviews/vaadin/flow/3912#-LANtpp2EWunh8GXC-18:-LAR_OTSBLlKf6_P3Uhy:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/41be9cf14089012a56c7c0c5fd792bbb45b50631/flow-server/src/main/java/com/vaadin/flow/internal/nodefeature/ElementListenerMap.java#L59)):*
<details><summary><i>Previously, SomeoneToIgnore (Kirill Bulatov) wrote…</i></summary><blockquote>

It's `null` already, but whatever :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3912)*
<!-- Sent from Reviewable.io -->

SomeoneToIgnore commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/flow/3912#-:-LARtv17C3btsW1VAKpb:bnfp4nl)*
<!-- Sent from Reviewable.io -->
